src/proc.cpp

```diff
@@ -189,17 +189,27 @@ bool job_t::signal(int signal) {
     return true;
 }
 
-statuses_t job_t::get_statuses() const {
+maybe_t<statuses_t> job_t::get_statuses() const {
     statuses_t st{};
+    bool has_status = false;
+    int laststatus = 0;
     st.pipestatus.reserve(processes.size());
     for (const auto &p : processes) {
         auto status = p->status;
+        if (status.is_empty()) {
+            st.pipestatus.push_back(laststatus);
```

Comment:
So a process that doesn't produce a status simply uses the status of the previous process in the pipeline?
For example `false | set x` will produce `1 1` (I understand this example is horribly contrived :laughing:)

I wonder if we can get away with dropping it, making pipestatus a singleton `1` in this example.
The user-visible `$pipestatus` will never be completely empty because it is only updated if at least one process in the job sets a status. I tried this and it seems to work well.

Reply:
The only reason I didn't do that was because I was worried that having pipestatus not match the number of `|` might be confusing. If you think that's not a big deal, I can make the change.

Reply:
Yeah, it's fine to leave it. I don't think anyone should ever rely on either behavior.
Maybe add a comment explaining in which case this branch is taken, because it is really quite an unusual scenario.

Reply:
Done